zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -20,21 +20,78 @@ use zebra_chain::{
 
 use DbFormatChange::*;
 
-use crate::{
-    constants::latest_version_for_adding_subtrees,
-    service::finalized_state::{DiskWriteBatch, ZebraDb},
-};
+use crate::service::finalized_state::ZebraDb;
 
 pub(crate) mod add_subtrees;
 pub(crate) mod cache_genesis_roots;
 pub(crate) mod fix_tree_key_type;
+pub(crate) mod no_migration;
+pub(crate) mod prune_trees;
+pub(crate) mod tree_keys_and_caches_upgrade;
 
 #[cfg(not(feature = "indexer"))]
 pub(crate) mod drop_tx_locs_by_spends;
 
 #[cfg(feature = "indexer")]
 pub(crate) mod track_tx_locs_by_spends;
 
+/// Defines method signature for running disk format upgrades.
+pub trait DiskFormatUpgrade {
+    /// Returns the version at which this upgrade is applied.
+    fn version(&self) -> Version;
+
+    /// Returns the description of this upgrade.
+    fn description(&self) -> &'static str;
+
+    /// Runs disk format upgrade.
+    fn run(
+        &self,
+        initial_tip_height: Height,
+        db: &ZebraDb,
+        cancel_receiver: &Receiver<CancelFormatChange>,
+    ) -> Result<(), CancelFormatChange>;
+
+    /// Check that state has been upgraded to this format correctly.
+    ///
+    /// # Panics
+    ///
+    /// If the state has not been upgraded to this format correctly.
```

Comment:
```suggestion
```

Reply:
+1

Reply:
Also would be good to explain the return value, e.g.

```
The outer Result indicates whether the validation was cancelled (due to e.g. node shutdown). The inner Result indicates whether the validation itself failed or not.
```